sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/DataTypeAstBuilder.scala

```diff
@@ -45,6 +46,41 @@ class DataTypeAstBuilder extends SqlBaseParserBaseVisitor[AnyRef] {
     withOrigin(ctx)(StructType(visitColTypeList(ctx.colTypeList)))
   }
 
+  override def visitSingleRoutineParamList(ctx: SingleRoutineParamListContext): StructType = {
+    withOrigin(ctx)(StructType(visitRoutineParamList(ctx.routineParamList())))
+  }
+
+  override def visitRoutineParamList(ctx: RoutineParamListContext): Seq[StructField] =
+    withOrigin(ctx) {
+      ctx.routineParam().asScala.map(visitRoutineParam).toSeq
+    }
+
+  override def visitRoutineParam(ctx: RoutineParamContext): StructField =
+    withOrigin(ctx) {
+      val builder = new MetadataBuilder
+      Option(ctx.defaultExpression()).foreach { expr =>
+        builder.putString("default", verifyAndGetExpression(expr.expression()))
+      }
+      val col = ctx.colType()
+      StructField(
+        name = col.colName.getText,
+        dataType = typedVisit[DataType](col.dataType()),
+        nullable = true,
+        metadata = builder.build())
+    }
+
+  private def verifyAndGetExpression(exprCtx: ExpressionContext): String = {
```

Comment:
did you copy this method from somewhere?

Reply:
refactored to avoid duplicating code with AstBuilder#getDefaultExpression